contributors/design-proposals/vertical-pod-autoscaler.md

```diff
@@ -0,0 +1,730 @@
+exceeding the request in a specific time window below a certain threshold**
+(e.g. below 1% in 24h). The window must be long (≥ 24h) to ensure that evictions
+caused by OOM do not visibly affect (a) availability of serving applications
+(b) progress of batch computations (a more advanced model could allow user to
+specify SLO to control this).
+
+#### Handling OOMs ####
+When a container is evicted due to exceeding available memory, its actual memory
+requirements are not known (the amount consumed obviously gives the lower
+bound). This is modelled by translating OOM events to artificial memory usage
+samples by applying a "safety margin" multiplier to the last observed usage.
+
+### History Storage ###
+VPA defines data access API for providers of historical events and resource
+utilization. Initially we will use Prometheus as the reference implementation of
+this API, at least for the resource utilization part. The historical events
+could be backed by another solution, e.g.
+[Infrastore](https://github.com/kubernetes/kubernetes/issues/44095).
+Users will be able to plug their own implementations.
+
+History Storage is populated with real time updates of resources utilization and
+events, similarly to the Recommender. The storage keeps at least 8 days of data.
+This data is only used to initialize the Recommender on startup.
+
+### Open questions ###
+1. How to resolve conflicts if multiple VPA objects match a Pod.
+
+2. How to adjust the recommendation before applying it to a specific pod,
+   based on the current state of the cluster (e.g. quota, space available on
+   nodes or other scheduling constraints).
+
+-----------
+Future work
+-----------
+
+### Pods that require VPA to start ###
+In the current proposal the Pod will be scheduled with originally configured
+resources if no matching VPA config is present at the Pod admission time.
+This may be undesired behavior. In particular the user may want to create the
+VPA config together with submitting the Pod, which leads to a race condition:
+the outcome depends on which resource (VPA or the Pod) is processed first.
+
+In order to address this problem we propose to allow marking Pods with a special
+annotation ("requires VPA") that prevents the Admission Controller from allowing
+the Pod if a corresponding VPA is not available.
+
+An alternative would be to introduce a VPA Initializer serving the same purpose.
+
+### Combining vertical and horizontal scaling ###
+In principle it may be possible to use both vertical and horizontal scaling for
+a single workload (group of Pods), as long as the two mechanisms operate on
+different resources.
+The right approach is to let the
+[Horizontal Pod Autoscaler](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/)
+scale the group based on the _bottleneck_ resource. The Vertical Pod Autoscaler
+could then control other resources. Examples:
+
+1. A CPU-bound workload can be scaled horizontally based on the CPU utilization
```

Comment:
I think it can be theoretically possible to use both HPA and VPA for scaling CPU-bound workloads. You can set a certain upper bound for CPU in VPA. When Pod's CPU is over this upper bound, HPA will kick in.